metals/src/main/scala/scala/meta/internal/bsp/ConnectionBspStatus.scala

```diff
@@ -13,25 +16,81 @@ class ConnectionBspStatus(
     bspStatus: BspStatus,
     folderPath: AbsolutePath,
     icons: Icons,
-) {
-  private val isServerResponsive = new AtomicBoolean(false)
-  val status: MetalsStatusParams => Unit = bspStatus.status(folderPath, _)
-
-  def connected(serverName: String): Unit =
-    if (isServerResponsive.compareAndSet(false, true))
-      status(ConnectionBspStatus.connectedParams(serverName, icons))
-  def noResponse(serverName: String): Unit =
-    if (isServerResponsive.compareAndSet(true, false)) {
-      scribe.debug("server liveness monitor detected no response")
-      status(ConnectionBspStatus.noResponseParams(serverName, icons))
-    }
+)(implicit rc: ReportContext) {
+  private val status = new AtomicReference[BspStatusState](
+    BspStatusState(Disconnected, None, None, shouldShow = false)
+  )
+  private val currentSessionErrors = new AtomicReference[Set[String]](Set())
+
+  def connected(serverName: String): Unit = changeState(Connected(serverName))
+
+  def noResponse(): Unit = {
+    scribe.debug("server liveness monitor detected no response")
+    changeState(NoResponse)
+  }
 
   def disconnected(): Unit = {
-    isServerResponsive.set(false)
-    status(ConnectionBspStatus.disconnectedParams)
+    currentSessionErrors.set(Set.empty)
+    changeState(Disconnected)
+  }
+
+  def showError(message: String, pathToReport: Path): Unit = {
+    val updatedSet =
+      currentSessionErrors.updateAndGet(_ + pathToReport.toUri().toString())
+    changeState(ErrorMessage(message), updatedSet)
+  }
+
+  def onReportsUpdate(): Unit = {
+    status.get().currentState match {
+      case ErrorMessage(_) => showState(status.get())
+      case _ =>
+    }
+  }
+
+  def isBuildServerResponsive: Option[Boolean] =
+    status.get().currentState match {
+      case NoResponse => Some(false)
+      case Disconnected => None
+      case _ => Some(true)
+    }
+
+  private def changeState(
+      newState: BspServerState,
+      errorReports: Set[String] = currentSessionErrors.get(),
+  ) = {
+    val newServerState = status.updateAndGet(_.changeState(newState))
+    if (newServerState.shouldShow) {
+      showState(newServerState, errorReports)
+    }
   }
 
-  def isBuildServerResponsive: Boolean = isServerResponsive.get()
+  private def showState(
+      statusState: BspStatusState,
+      errorReports: Set[String] = currentSessionErrors.get(),
+  ) = {
+    val showParams =
+      statusState.currentState match {
+        case Disconnected => ConnectionBspStatus.disconnectedParams
+        case NoResponse =>
+          ConnectionBspStatus.noResponseParams(statusState.serverName, icons)
+        case Connected(serverName) =>
+          ConnectionBspStatus.connectedParams(serverName, icons)
+        case ErrorMessage(message) =>
+          val currentSessionReports = errorReports.intersect(
```

Comment:
Deleting or renaming the report is a way to say, "I saw that report" so it disappears from the status (you can see that on the gif). Otherwise I think it may be annoying for people to see that warning even if it's not relevant. Reports also take care of deduplication. So if you have a bloop error report `report_10am` and the error happens again at 3pm, the report gets moved to file `report_3pm`. So now when you sync you only have one error: `report_3pm`.

The drawback is that you can't really "mute" an error, you can make the warning disappear but it will show up again when the error happens. We could use special suffixes to reports, e.g. `mute` to mute a report. Maybe it gets a bit hairy but I gives the user a pretty good control over the messages without any special API that the clients need to implement.

Reply:
ok, thanks for explaining, could you add a bit of that in a comment?